File: test/cubcav/parti.pl
$nsubdo = 30;
for ($s = 0; $s<$nsubdo; $s++) {
    $nelem = count_lines("icone$s.dat");
    print <<EOF;
#================================================================
object "s$s" class array type int rank 1 shape 4
	items $nelem data file "icone$s.dat"
attribute "element type" string "tetrahedra"
attribute "ref" string "positions"

object "colors$s" class array type float rank 1 shape 3 items $nelem
           data file "colors$s.dat"
attribute "dep" string "connections"

object "subdo$s" class field 
component "positions" value "xnod"
component "connections" value "s$s"
component "colors" value "colors$s"

EOF
}

$ngroup = 6;
$sg = ceil($nsubdo/$ngroup);
for ($g = 0; $g<$ngroup; $g++) {
    $s1 = $g*$sg;
    $s2 = $s1+$sg;
    $s2 = $nsubdo if $s2>$nsubdo;
    print "object \"sg$g\" class group\n";
    for ($j=$s1; $j<$s2; $j++) {
	print "member \"subdo$j\" value \"subdo$j\"\n";
    }
    print "\n";
}

print "object \"group\" class group\n";
for ($g = 0; $g<$ngroup; $g++) {
    print "member \"sg$g\" value \"sg$g\"\n";
}
print "\n";
print "end\n";

1;

File: test/cubcav/parti.m
###key parti.m

## Reads a node partitioning and the connectivity and
## computes an element partitioning for visualization with DX

if 1
  npart=aload("node-part.dat");
  npart=npart(:,2);
  icone=aload("cubcav.con-tetra.tmp");
  xnod = aload("cubcav.nod.tmp");
endif

icosub=icone;
nelem=rows(icosub);
for k=1:columns(icosub)
  icosub(:,k) = npart(icosub(:,k));
endfor

nsubdo = max(npart)+1;
subdo = -ones(nelem,1);
vert = zeros(nelem,1);
for k=0:nsubdo-1
  vert1 = sum(icosub'==k)';
  subdo = choose(vert<vert1,k*ones(nelem,1),subdo);
endfor

for k=0:nsubdo-1
  printf("in subd %d: %d\n",k,sum(subdo==k));
endfor

nnod = rows(npart);
in_subdo = zeros(nnod,nsubdo);
asave("epart.dat",subdo);

xele=pfnd2ele(xnod,icone,xnod);

colors = [1 0 0;
	  0 1 0;
	  1 1 0;
	  0 1 1;
	  1 0 1];
zsubdo=zeros(nsubdo,1);

for k=0:nsubdo-1
  indx = find(subdo==k);
  zsubdo(k+1) = mean(xele(indx,3));
endfor

[bid,sbdindx] = sort(-zsubdo);

for k=1:nsubdo
  indx = find(subdo==sbdindx(k)-1);
  color=colors(rem(k,rows(colors))+1,:);
  color=color(ones(rows(indx),1),:);
  eval(["asave(\"icone" int2str(k-1) ".dat\",icone(indx,:)-1);"]);
  eval(["asave(\"colors" int2str(k-1) ".dat\",color);"]);
endfor
